core/server/master/src/main/java/alluxio/master/file/meta/InvalidationSyncCache.java

```diff
@@ -0,0 +1,418 @@
+        // we always check if the current path has been invalidated
+        lastInvalidationTime = Math.max(lastInvalidationTime,
+            syncState.mInvalidationTime);
+        switch (parentLevel) {
+          case 0: // the base path
+            isFile = syncState.mIsFile;
+            switch (descendantType) {
+              case NONE:
+                // we are syncing no children, so we use our validation time
+                lastValidationTime = Math.max(lastValidationTime, syncState.mValidationTime);
+                break;
+              case ONE:
+                // if the last sync on this path was a file, then we can use
+                // the normal validation time, as it has no children
+                lastValidationTime = Math.max(lastValidationTime, isFile
+                    ? syncState.mValidationTime : syncState.mDirectValidationTime);
+                // since we are syncing the children, we must check if a child was invalidated
+                lastInvalidationTime = Math.max(lastInvalidationTime,
+                    syncState.mDirectChildInvalidation);
+                break;
+              case ALL:
+                // if the last sync on this path was a file, then we can use
+                // the normal validation time, as it has no children
+                lastValidationTime = Math.max(lastValidationTime, isFile
+                    ? syncState.mValidationTime : syncState.mRecursiveValidationTime);
+                // since we are syncing recursively, we must check if any recursive
+                // child was invalidated
+                lastInvalidationTime = Math.max(lastInvalidationTime,
+                    syncState.mRecursiveChildInvalidation);
+                break;
+              default:
+                throw new RuntimeException("Unexpected descendant type " + descendantType);
+            }
+            break;
+          case 1: // at the parent path
+            lastValidationTime = Math.max(lastValidationTime,
+                isFile ? syncState.mDirectValidationTime :
+                descendantType != DescendantType.NONE ? syncState.mRecursiveValidationTime
+                    : syncState.mDirectValidationTime);
+            break;
+          default: // at a higher ancestor path
+            lastValidationTime = Math.max(lastValidationTime, syncState.mRecursiveValidationTime);
+        }
+      }
+      if (currPath.equals(AlluxioURI.SEPARATOR)) {
+        break;
+      }
+      currPath = PathUtils.getParent(currPath);
+      parentLevel++;
+    }
+    long currentTime = mClock.millis();
+    SyncCheck result;
+
+    if (intervalMs == 0) {
+      // Always sync.
+      return SyncCheck.shouldSyncWithTime(lastValidationTime);
+    }
+    if (lastInvalidationTime == 0 && intervalMs < 0) {
+      // if no invalidation, and a negative interval then do not sync
+      result = SyncCheck.shouldNotSyncWithTime(lastValidationTime);
+    } else if (lastInvalidationTime >= lastValidationTime) {
```

Comment:
can this be `lastInvalidationTime=0` and `lastValidationTime=0`?

Reply:
Yes this would indicate the first time this path is being synced, so it will perform a sync.